distributed/shuffle/_core.py

```diff
@@ -208,15 +208,33 @@ def _get_assigned_worker(self, i: _T_partition_id) -> str:
     async def _receive(self, data: list[tuple[_T_partition_id, bytes]]) -> None:
         """Receive shards belonging to output partitions of this shuffle run"""
 
-    @abc.abstractmethod
     async def add_partition(
         self, data: _T_partition_type, partition_id: _T_partition_id
     ) -> int:
-        """Add an input partition to the shuffle run"""
+        self.raise_if_closed()
+        if self.transferred:
+            raise RuntimeError(f"Cannot add more partitions to {self}")
+        return await self._add_partition(data, partition_id)
 
     @abc.abstractmethod
+    async def _add_partition(
+        self, data: _T_partition_type, partition_id: _T_partition_id
+    ) -> int:
+        """Add an input partition to the shuffle run"""
+
     async def get_output_partition(
         self, partition_id: _T_partition_id, key: str, **kwargs: Any
+    ) -> _T_partition_type:
+        self.raise_if_closed()
+        await self._ensure_output_worker(partition_id, key)
+        if not self.transferred:
+            raise RuntimeError("`get_output_partition` called before barrier task")
```

Comment:
@fjetter: Switching these two checks around fixed the problem @phofl saw in testing.

* [ ] Add a deterministic test